### PR TITLE
Fix compilation on Windows when building with CUDA.

### LINKED ISSUE #2019: [windows]Compiling from source
I have checked out the source code of Open3D **v0.10.0**.  Then used CMake to configure and generate a VS solution. The compilation environment as follows.

- Operating System: Windows 10(1909)
- IDE: Visual Studio 2017 Community
- Open3D: 0.10.0
- CMake: 3.17.3
- CUDA Toolkit SDK: 10.0

### Problem 1:
If I checked **BUILD_BENCHMARKS** on the CMake gui, following codes in the KDTreeFlann.cpp compiled failure.
```
static void BM_TestKDTreeLine0(benchmark::State& state) {
    // state.range(n) are arguments that are passed to us
    int radius = state.range(0);
    int size = state.range(1);
    testKDTreeLine0.setup(size);
    for (auto _ : state) {
        testKDTreeLine0.search(radius);
    }
}
```
image

**How to slove?**
I need to manually make a explicitly cast to the variables _radius_ and _size_ to make it well.

### Problem 2:
If I checked **BUILD_CUDA_MODULE** on the CMake gui, then rebuild the **ALL_BUILD** target. More compilation problems turn out. The version of CUDA Toolkit SDK is v10.0.
```
17>F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\build\src\Open3D\Core>"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.0\bin\nvcc.exe" -gencode=arch=compute_30,code=\"sm_30,compute_30\" -gencode=arch=compute_35,code=\"sm_35,compute_35\" -gencode=arch=compute_50,code=\"sm_50,compute_50\" -gencode=arch=compute_52,code=\"sm_52,compute_52\" -gencode=arch=compute_60,code=\"sm_60,compute_60\" -gencode=arch=compute_61,code=\"sm_61,compute_61\" -gencode=arch=compute_70,code=\"sm_70,compute_70\" -gencode=arch=compute_75,code=\"sm_75,compute_75\" -gencode=arch=compute_70,code=\"compute_70,compute_70\" -gencode=arch=compute_75,code=\"compute_75,compute_75\" --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX86\x64" -x cu  -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\build\3rdparty_install\include" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\src" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.0\include" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\dirent" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\dirent\dirent" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\Eigen" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\flann" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\flann\flann" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\flann\flann\algorithms" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\flann\flann\nn" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\flann\flann\util" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\glew\include" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\GLFW\include" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\jsoncpp\include" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\liblzf" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\liblzf\liblzf" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\tomasakeninemoeller\include" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\tomasakeninemoeller\include\tomasakeninemoeller" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\libpng" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\rply" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\rply\rply" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\rply\rply\etc" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\rply\rply\manual" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\tinyfiledialogs\include" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\tinygltf" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\tinyobjloader" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\qhull\src" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\fmt\include" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\PoissonRecon" -I"F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\3rdparty\librealsense\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.0\include"     --keep-dir x64\Release -maxrregcount=0  --machine 64 --compile -cudart static --expt-extended-lambda /MP -Xcompiler="/EHsc -Ob2"    -D_WINDOWS -D_GLIBCXX_USE_CXX11_ABI=0 -DNDEBUG -DBUILD_CUDA_MODULE -DWINDOWS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -DGLEW_STATIC -DNOMINMAX -D_USE_MATH_DEFINES -D"CMAKE_INTDIR=\"Release\"" -DWIN32 -D_WINDOWS -D_GLIBCXX_USE_CXX11_ABI=0 -D_ENABLE_EXTENDED_ALIGNED_STORAGE -DNDEBUG -DBUILD_CUDA_MODULE -DWINDOWS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -DGLEW_STATIC -DNOMINMAX -D_USE_MATH_DEFINES -D"CMAKE_INTDIR=\"Release\"" -D_MBCS -Xcompiler "/EHsc /W3 /nologo /O2 /FdCore.dir\Release\Core.pdb /FS /Zi  /MD /GR" -o Core.dir\Release\UnaryEWCUDA.obj "F:\04_Software\3rdParty\Open3D\release\Open3D-0.10.0\src\Open3D\Core\Kernel\UnaryEWCUDA.cu"
17>nvcc fatal   : A single input file is required for a non-link phase when an outputfile is specified
17>
```

I don't know how to deal with it. Does anyone could help me?

### LINKED ISSUE #3158: Build error when building from source (Windows 10 64 - Visual Studio 19)
**IMPORTANT: Please use the following template to report the bug.**
_____

**Describe the bug**
Building Open3D from source leaves me with 55 errors when building the ALL_BUILD target. See build output and steps below. 
Most probably the error sits in front of the screen, however maybe someone can point me in the right direction. 

**To Reproduce**

1. Clone Open3d repository into new folder using git bash `git clone --recursive https://github.com/intel-isl/Open3D` 
2. Open CMake GUI (v3.18.4) with the following Configuration/Changes:
     a) Where is the source code: `C:/Users/Johannes/source/repos/Open3D/Open3D`
     b) Where to build the binaries: `C:/Users/Johannes/source/repos/Open3D/Open3D/build` 
     c) CMAKE_INSTALL_PREFIX: `C:/Open3D`
     d) PYTHON_EXECUTABLE: `C:\Users\Johannes\Anaconda3\python.exe`
     e) PYTHON_IN_PATH: `C:\Users\Johannes\Anaconda3\python.exe`
     d) BUILD_CUDA_MODULE: `TRUE`
3. Configure for Visual Studio 2019 until no errors are shown, none of the configuration is marked red
4. Generate
5. Open Project
6. Change Build Configuration to "Release"
7. Right-Click CMakePredefinedTargets/ALL_BUILD and select `Build` (or `Rebuild`)

**Expected behavior**
Except for building with CUDA Modules this is mostly the standard procedure described in Build from [Source Section of the documentation](http://www.open3d.org/docs/release/compilation.html#). CMAKE doesn't show any errors and finds all the path to the installed NVIDIA CUDA Toolkit (v11.2) items as well as the nvcc compiler. Not sure, why the build fails. 

**Build Output**
[build_output.txt](https://github.com/intel-isl/Open3D/files/6115817/build_output.txt)

**Error List**
[error_list.txt](https://github.com/intel-isl/Open3D/files/6115818/error_list.txt)

**Environment (please complete the following information):**

- Operating system: Windows 10 64-bit
- Python version: Python 3.7.4 (Anaconda)
- Open3D version: 0.12.0
- Is this remote workstation?: no
- How did you install Open3D?: build from source
- Compiler version (if built from source): VS2019

### PR DESCRIPTION
**Pull request by Dudulle**
There were a few different issues during build. Mainly the following:

- A few compiler options were added to nvcc instead of the c++ compiler.
- A few warnings treated as errors. Some were fixed and others were silenced.
- Member variables were not initialized in the right order in some parts of CUDA. The reorder check was removed although there might be a better fix for this?
- nvcc does not support the [[deprecated]] attribute which is used in fmt.

This was tested with CUDA 10.1, 10.2, and 11.2.

Fixes #2019 (other issues mentioned already fixed in master branch)
Fixes #3158  (other issue mentioned fixed in master branch)

---
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3036)

### CODE REVIEW

File: 3rdparty/find_dependencies.cmake [hunk 1]
```diff
@@ -1070,16 +1070,16 @@ else()
...
         target_link_libraries(3rdparty_mkl INTERFACE Threads::Threads ${CMAKE_DL_LIBS})
-        # cuSOLVER and cuBLAS
-        if(BUILD_CUDA_MODULE)
-            target_link_libraries(3rdparty_mkl INTERFACE
-                                ${CUDA_cusolver_LIBRARY}
-                                ${CUDA_CUBLAS_LIBRARIES})
-        endif()
     elseif(MSVC)
-        target_compile_options(3rdparty_mkl INTERFACE "/DMKL_ILP64")
+        target_compile_options(3rdparty_mkl INTERFACE "$<$<COMPILE_LANGUAGE:CXX>:/DMKL_ILP64>")
```
Comment (ssheorey) [inline 3rdparty/find_dependencies.cmake:1076]: `target_compile_definitions(3rdparty_mkl INTERFACE "$<$<COMPILE_LANGUAGE:CXX>:MKL_ILP64>")`
can be common to both `UNIX` and `MSVC`

**Comment by ssheorey** [comment]
Thanks for this PR. This has been a long standing user request. We don't have a CI setup to test Windows builds with CUDA currently. Were you able to run all C++ and Python unit tests with this CUDA build? Please let us know if any tests are failing with CUDA.

**Comment by Dudulle** [comment]
@ssheorey I only ran the c++ tests for now and will try the python tests in the next few days.

In the c++ tests, 5 related to cuda failed with an out of memory error. I did not look into it much yet but it seems like ~2GB of memory (my card has 4GB) is being allocated and not deallocated in a passing test.

**Comment by ssheorey** [comment]
@Dudulle Can you paste the results of the unit tests here? We can take a hand at fixing things if something looks broken.

**Comment by Dudulle** [comment]
@ssheorey Here are the results of the c++ unit tests:
> [----------] Global test environment tear-down
[==========] 1131 tests from 74 test suites ran. (86062 ms total)
[  PASSED  ] 1120 tests.
[  SKIPPED ] 1 test, listed below:
[  SKIPPED ] RGBDImage.CreateFromNYUFormat
[  FAILED  ] 10 tests, listed below:
[  FAILED  ] FileSystem.File_Exists_Remove
[  FAILED  ] FileSystem.ListFilesInDirectory
[  FAILED  ] FileSystem.ListFilesInDirectoryWithExtension
[  FAILED  ] FileSystem.GetPathComponents
[  FAILED  ] Tensor/TensorPermuteDevices.ReduceSum64bit1D/1, where GetParam() = 8-byte object <01-00 00-00 00-00 00-00>
[  FAILED  ] Tensor/TensorPermuteDevices.ReduceSum64bit2DCase0/1, where GetParam() = 8-byte object <01-00 00-00 00-00 00-00>
[  FAILED  ] Tensor/TensorPermuteDevices.ReduceSum64bit2DCase1/1, where GetParam() = 8-byte object <01-00 00-00 00-00 00-00>
[  FAILED  ] Tensor/TensorPermuteDevices.ReduceSum64bit2DCase2/1, where GetParam() = 8-byte object <01-00 00-00 00-00 00-00>
[  FAILED  ] Tensor/TensorPermuteDevices.ReduceSum64bit2DCase3/1, where GetParam() = 8-byte object <01-00 00-00 00-00 00-00>
[  FAILED  ] Tensor/TensorPermuteDevices.ReduceSumLargeArray/1, where GetParam() = 8-byte object <01-00 00-00 00-00 00-00>

I've looked a bit deeper and no memory seems to be released in the Tensor tests.
The big chunk of memory is allocated here but it seems it is not the only place where there is a leak:
https://github.com/intel-isl/Open3D/blob/31e5d4d16eabbe05d5bb2765deaa7c13c4dcd18a/cpp/tests/core/Tensor.cpp#L1562
I'm also reaching this line: https://github.com/intel-isl/Open3D/blob/31e5d4d16eabbe05d5bb2765deaa7c13c4dcd18a/cpp/open3d/core/MemoryManagerCUDACached.cu#L121

The python package also does not work so I could not run the python tests.
I don't use python much so I'll leave that part for someone with more experience.
First there is the following error:
> Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "L:\Source\Open3D\.venv38\lib\site-packages\open3d\__init__.py", line 62, in <module>
    _pybind_cuda = _CDLL(
  File "C:\Users\User\AppData\Local\Programs\Python\Python38\lib\ctypes\__init__.py", line 363, in __init__
    if '/' in name or '\\' in name:
TypeError: argument of type 'WindowsPath' is not iterable

If I use hardcoded paths it then says it cannot find the module or a dependency.

**Review by yxlao — CHANGES_REQUESTED** [review]
Looks good to me in general. Regarding the silenced MSVC warnings, we can create a separate PR to actually fix the warnings and re-enable the flags.

Could you make some additional changes:
1. merge master (include changes from #3070)
2. add the following to `.github/workflows/windows.yml` to run unit test on Windows CI. This only runs the CPU unit tests, but it is already a substantial improvement. 
    ```diff
    diff --git a/.github/workflows/windows.yml b/.github/workflows/windows.yml
    index 626989dbf..f03f4b38a 100644
    --- a/.github/workflows/windows.yml
    +++ b/.github/workflows/windows.yml
    @@ -75,6 +75,10 @@ jobs:
                 --target build-examples-iteratively
               cmake --build . --parallel ${{ env.NPROC }} --config ${{ matrix.CONFIG }} `
                 --target INSTALL
    +      - name: Run C++ unit tests
    +        working-directory: ${{ env.BUILD_DIR }}
    +        run: |
    +          .\bin\${{ matrix.CONFIG }}\tests.exe
           - name: Linking to Open3D
             run: |
               $ErrorActionPreference = 'Stop'
    ```
3. For the failed tests, exclude them with  `gtest_filter`. E.g. `--gtest_filter="-*Reduce*Sum*"` will filter out the big `Sum` unit tests that can cause the out-of-memory error. See https://stackoverflow.com/a/14021997/1255535 for details.

**Comment by Dudulle** [comment]
@yxlao Thanks for the feedback.

I've added the change to enable the unit tests on CI.
I've also pushed a few changes I made yesterday to remove warning C4456 and a couple assertions I ran into.

For the remaining warnings silenced here, besides the type conversion warnings all of them come from either CUDA or Eigen.
I've noticed the warnings in Eigen are all fixed upstream since end of January but one is added [here](https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/CwiseBinaryOp.h#L106), something about using `__host__` and `__device__` before a default constructor. I did not update it in this PR as the current version is quite old so I'm unsure it is safe to do so.

**Comment by Dudulle** [comment]
It seems a few CI are out of memory during the tests.
On my side I now have one test failing (with CUDA, works on CPU) from the new tests in ba9200b3915bacf22cab002024205adfe92d7492
`[  FAILED  ] Image/ImagePermuteDevices.FilterSobel/1, where GetParam() = 8-byte object <01-00 00-00 00-00 00-00>`
One of the CI seems to fail another of these tests and another one in UniformTSDFVolume.RealData.
And some fail to build with unresolved symbol in ippiw.lib.

**Comment by yxlao** [comment]
@Dudulle Right, it is fine that some tests are not passing just now. You may add `--gtest_filter` to filter out those tests in `windows.yml` for now. The tests can be fixed in later PRs.

**Comment by Dudulle** [comment]
@yxlao Thanks,
Besides the failing tests, the builds with BUILD_SHARED_LIBS=ON will all fail because of an assert ([know issue apparently](https://github.com/zeromq/czmq/issues/1788)):
`Assertion failed: Successful WSASTARTUP not yet performed [10093] (C:\Open3D\build\zeromq\src\ext_zeromq\src\signaler.cpp:190)`
And the remaining ones with STATIC_WINDOWS_RUNTIME=ON fail to build the tests.

Should I change it to only build the tests for when they are both off and disable the tests going out of memory?

**Comment by yxlao** [comment]
> Should I change it to only build the tests for when they are both off and disable the tests going out of memory?

Yes.

To disable a GitHub actions step, we can do something similar to this:
https://github.com/intel-isl/Open3D/blob/f65a6f10c4683cc6885fe6e7c0762365aff7efbc/.github/workflows/windows.yml#L102

And it seems that we can `&&` the two conditions together (see [here](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#operators)).

**Review by yxlao — APPROVED** [review]

**Comment by ssheorey** [comment]
I'm looking into this error:
> Traceback (most recent call last):
> File "", line 1, in
> File "L:\Source\Open3D.venv38\lib\site-packages\open3d_init_.py", line 62, in
> _pybind_cuda = CDLL(
> File "C:\Users\User\AppData\Local\Programs\Python\Python38\lib\ctypes_init.py", line 363, in init
> if '/' in name or '\' in name:
> TypeError: argument of type 'WindowsPath' is not iterable

**Comment by Dudulle** [comment]
Rebased on master to resolve conflicts in `Tensor.h`

**Review by Dudulle — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3036)* status: 0 of 13 files reviewed, 1 unresolved discussion (waiting on @ssheorey)

___
*[3rdparty/find_dependencies.cmake, line 1076 at r1](https://reviewable.io/reviews/intel-isl/open3d/3036#-MTSYa5KHJy_SNdf44p3-r1-1076:-MUwsvTq1GgfKvE-JmqV:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/5ffff22cab8a15af7328e7cf72ac846d0d8c25b1/3rdparty/find_dependencies.cmake#L1076)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

`target_compile_definitions(3rdparty_mkl INTERFACE "$<$<COMPILE_LANGUAGE:CXX>:MKL_ILP64>")`
can be common to both `UNIX` and `MSVC`
</blockquote></details>

Done.

**Comment by yxlao** [comment]
There are various topics here. Let's summarize them. 

Here's my current understanding: 

1. Windows CUDA compilation and running basic things with C++: supported in this PR
2. Windows C++ unit test CPU: supported for `SHARED: OFF, STATIC_RUNTIME: OFF`, with some tests filtered
3. Windows C++ unit test CUDA: what is the status?
4. Windows CUDA compilation and running basic things with Python wheel: what is the status?
5. Windows Python unit test CPU: what is the status?
6. Windows Python unit test CUDA: what is the status?

@Dudulle could you correct me on the table above? It is just for documenting what is expected to work, we don't need to address all of them in this PR,

@ssheorey does the error `TypeError: argument of type 'WindowsPath' is not iterable` belong to item 4. above?

**Comment by Dudulle** [comment]
@yxlao

> Windows CUDA compilation and running basic things with C++: supported in this PR
Windows C++ unit test CPU: supported for SHARED: OFF, STATIC_RUNTIME: OFF, with some tests filtered

Correct. Filtered tests fail because of lack of memory.
Tests would pass with `SHARED_LIBS: ON` and `STATIC_RUNTIME: OFF` but fail because of an assert in `zeromq` on exit.
Linking error with `SHARED_LIBS: OFF` and `STATIC_RUNTIME: ON` (unresolved symbols in `ippiw.lib`)

>Windows C++ unit test CUDA: what is the status?

All tests pass except `Image/ImagePermuteDevices.FilterSobel`.
Tests disabled in CI also fail for me on CUDA because of lack of memory.
Tested with `SHARED_LIBS: OFF` and `STATIC_RUNTIME: OFF`.

>Windows CUDA compilation and running basic things with Python wheel: what is the status?
Windows Python unit test CUDA: what is the status?

Python module compiles and can be installed but fails when trying to import.

>Windows Python unit test CPU: what is the status?

Python CPU tests pass.

**Comment by ssheorey** [comment]
> There are various topics here. Let's summarize them.
> 
> Here's my current understanding:
>     1. Windows CUDA compilation and running basic things with C++: supported in this PR
>     2. Windows C++ unit test CPU: supported for `SHARED: OFF, STATIC_RUNTIME: OFF`, with some tests filtered

It should be OK to turn off the filtering. All tests pass with `SHARED: OFF, STATIC_RUNTIME: OFF` on an 8 vCPU / 30GB / Tesla T4 VM:

```
> PS C:\Users\sameer_sheorey\Open3D\Open3D\build> .\bin\RelWithDebInfo\tests.exe
> [==========] Running 886 tests from 72 test suites.
> [----------] Global test environment set-up.
> [----------] 1 test from PinholeCameraTrajectory
> [ RUN      ] PinholeCameraTrajectory.ConvertToFromJsonValue
> [       OK ] PinholeCameraTrajectory.ConvertToFromJsonValue (1 ms)
> [----------] 1 test from PinholeCameraTrajectory (1 ms total)
> ...
> ...
> [----------] 1 test from Image/ImagePermuteDevicePairs
> [ RUN      ] Image/ImagePermuteDevicePairs.CopyDevice/0
> [       OK ] Image/ImagePermuteDevicePairs.CopyDevice/0 (0 ms)
> [----------] 1 test from Image/ImagePermuteDevicePairs (0 ms total)
> 
> [----------] Global test environment tear-down
> [==========] 886 tests from 72 test suites ran. (88143 ms total)
> [  PASSED  ] 885 tests.
> [  SKIPPED ] 1 test, listed below:
> [  SKIPPED ] RGBDImage.CreateFromNYUFormat
> 
>   YOU HAVE 210 DISABLED TESTS
> 
```

>     3. Windows C++ unit test CUDA: what is the status?

All tests pass:
```
> PS C:\Users\sameer_sheorey\Open3D\Open3D\build> .\bin\RelWithDebInfo\tests.exe
> [==========] Running 1143 tests from 74 test suites.
> [----------] Global test environment set-up.
> [----------] 1 test from PinholeCameraTrajectory
> [ RUN      ] PinholeCameraTrajectory.ConvertToFromJsonValue
> [       OK ] PinholeCameraTrajectory.ConvertToFromJsonValue (1 ms)
> [----------] 1 test from PinholeCameraTrajectory (2 ms total)
> ...
> 
> [----------] 4 tests from Registration/RegistrationPermuteDevices
> [ RUN      ] Registration/RegistrationPermuteDevices.ICPConvergenceCriteriaConstructor/0
> [       OK ] Registration/RegistrationPermuteDevices.ICPConvergenceCriteriaConstructor/0 (0 ms)
> [ RUN      ] Registration/RegistrationPermuteDevices.ICPConvergenceCriteriaConstructor/1
> [       OK ] Registration/RegistrationPermuteDevices.ICPConvergenceCriteriaConstructor/1 (1 ms)
> [ RUN      ] Registration/RegistrationPermuteDevices.RegistrationResultConstructor/0
> [       OK ] Registration/RegistrationPermuteDevices.RegistrationResultConstructor/0 (0 ms)
> [ RUN      ] Registration/RegistrationPermuteDevices.RegistrationResultConstructor/1
> [       OK ] Registration/RegistrationPermuteDevices.RegistrationResultConstructor/1 (0 ms)
> [----------] 4 tests from Registration/RegistrationPermuteDevices (2 ms total)
> 
> ...
> [----------] Global test environment tear-down
> [==========] 1143 tests from 74 test suites ran. (129320 ms total)
> [  PASSED  ] 1142 tests.
> [  SKIPPED ] 1 test, listed below:
> [  SKIPPED ] RGBDImage.CreateFromNYUFormat


  YOU HAVE 210 DISABLED TESTS
```
>     4. Windows CUDA compilation and running basic things with Python wheel: what is the status?
>     5. Windows Python unit test CPU: what is the status?

The attached patch fixes the Python issues. All Python tests pass. I added WITH_FAISS to `open3d._build_config` to skip 2 Python tests if we don't build with FAISS.
[0001-Windows-CUDA-Python-support.patch.zip](https://github.com/intel-isl/Open3D/files/6094301/0001-Windows-CUDA-Python-support.patch.zip)

```
> PS C:\Users\sameer_sheorey\Open3D\Open3D\python\test> python -m pytest --ignore ml --ignore ml_ops
> ================================================= test session starts =================================================
> platform win32 -- Python 3.6.8, pytest-6.2.1, py-1.10.0, pluggy-0.13.1
> rootdir: C:\Users\sameer_sheorey\Open3D\Open3D\python
> collected 331 items
> 
> test_color_map.py .                                                                                              [  0%]
> test_flags.py .                                                                                                  [  0%]
> test_octree.py .........                                                                                         [  3%]
> test_open3d_eigen.py .......................................                                                     [ 15%]
> test_remote_functions.py s                                                                                       [ 15%]
> core\test_core.py .............................................................................................. [ 43%]
> ................................................................................................................ [ 77%]
> .................................                                                                                [ 87%]
> core\test_hashmap.py ......                                                                                      [ 89%]
> core\test_linalg.py ................                                                                             [ 94%]
> core\test_nn.py ss......                                                                                         [ 96%]
> core\test_torch_integration.py ...                                                                               [ 97%]
> t\geometry\test_pointcloud.py ....                                                                               [ 98%]
> t\geometry\test_tensormap.py .                                                                                   [ 99%]
> t\geometry\test_voxelgrid.py .                                                                                   [ 99%]
> t\io\test_realsense.py ss                                                                                        [100%]
> 
> =========================================== 326 passed, 5 skipped in 39.54s ===========================================
```

>     6. Windows Python unit test CUDA: what is the status?

```

> PS C:\Users\sameer_sheorey\Open3D\Open3D\python\test> python -m pytest --ignore ml --ignore ml_ops
> ======================================================== test session starts =========================================================
> platform win32 -- Python 3.6.8, pytest-6.2.1, py-1.10.0, pluggy-0.13.1
> rootdir: C:\Users\sameer_sheorey\Open3D\Open3D\python
> collected 599 items
> 
> test_color_map.py .                                                                                                             [  0%]
> test_flags.py .                                                                                                                 [  0%]
> test_octree.py .........                                                                                                        [  1%]
> test_open3d_eigen.py .......................................                                                                    [  8%]
> test_remote_functions.py s                                                                                                      [  8%]
> core\test_core.py ............................................................................................................. [ 26%]
> ............................................................................................................................... [ 47%]
> ............................................................................................................................... [ 69%]
> ..............................................................................................................                  [ 87%]
> core\test_hashmap.py ............                                                                                               [ 89%]
> core\test_linalg.py ................................                                                                            [ 94%]
> core\test_nn.py ssss........                                                                                                    [ 96%]
> core\test_torch_integration.py .....                                                                                            [ 97%]
> t\geometry\test_pointcloud.py ........                                                                                          [ 98%]
> t\geometry\test_tensormap.py ..                                                                                                 [ 99%]
> t\geometry\test_voxelgrid.py ..                                                                                                 [ 99%]
> t\io\test_realsense.py ss                                                                                                       [100%]
> 
> ================================================== 592 passed, 7 skipped in 40.76s ===================================================

```

 
> @Dudulle could you correct me on the table above? It is just for documenting what is expected to work, we don't need to address all of them in this PR,
> 
> @ssheorey does the error `TypeError: argument of type 'WindowsPath' is not iterable` belong to item 4. above?

Yes, that is fixed in this patch.

**Comment by ssheorey** [comment]
@Dudulle Can you try the patch? If it work for you, go ahead and commit it.

**Comment by Dudulle** [comment]
@ssheorey I have tried your patch but I am still having the following error:

>File "<string>", line 1, in <module>
  File "L:\Source\Open3D\.venv38\lib\site-packages\open3d\__init__.py", line 62, in <module>
    _pybind_cuda = _CDLL(
  File "C:\Users\Cyril\AppData\Local\Programs\Python\Python38\lib\ctypes\__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
FileNotFoundError: Could not find module 'L:\Source\Open3D\.venv38\lib\site-packages\open3d\cuda\pybind.cp38-win_amd64.pyd' (or one of its dependencies). Try using the full path with constructor syntax.

Maybe this is because of my configuration. Should I enable something else than `BUILD_PYTHON_MODULE`?

**Comment by Dudulle** [comment]
Added an extra commit to support Visual Studio 16.9+

**Comment by Dudulle** [comment]
Rebased on master to resolve conflicts in `CMakeLists.txt`.
Enabled all unit tests except for `SHARED=ON,STATIC_RUNTIME=ON` configuration.

**Comment by yxlao** [comment]
thanks @Dudulle

**Pull request merged**